### PR TITLE
Add .diff and .patch support to compare

### PR DESCRIPTION
**Pull request by badhezi**
Implements proposal https://github.com/go-gitea/gitea/issues/34410
Screenshot 2025-05-12 at 17 25 44
Screenshot 2025-05-12 at 17 25 37

### CODE REVIEW

**Comment by wxiaoguang** [comment]
Some things in my mind:

1. It needs some tests to cover the behavior, otherwise this feature would get broken in the future.
2. It needs to handle edge cases like `the.branch.name.contains.diff`, the logic should be like this:
    * Parse `{ref1}...{ref2}`
    * If `ref2` exists, then do not trim any suffix
    * If `ref2` doesn't exist and it ends with `.diff` or `.patch`, then respond the diff/patch
3. Maybe it could add some dropdown menu items on the UI to make users could download the diff/patch by some clicks

File: modules/git/tree.go [hunk 1]
```diff
@@ -73,3 +73,14 @@ func (repo *Repository) GetTreePathLatestCommit(refName, treePath string) (*Comm
 	}
 	return repo.GetCommit(strings.TrimSpace(stdout))
 }
+
+// RevParse resolves a revision reference to other git-related objects
+func (repo *Repository) RevParse(ref, file string) (string, error) {
```
Comment (wxiaoguang) [inline modules/git/tree.go]: Where is it really used except the tests?

File: routers/web/repo/compare.go [hunk 2]
```diff
@@ -228,10 +228,31 @@ func ParseCompareInfo(ctx *context.Context) *common.CompareInfo {
...
 		infos = []string{baseRepo.DefaultBranch, baseRepo.DefaultBranch}
 	} else {
+		// check if head is a branch or tag only if infoPath ends with .diff or .patch
+		if strings.HasSuffix(infoPath, ".diff") || strings.HasSuffix(infoPath, ".patch") {
+			infos = strings.SplitN(infoPath, "...", 2)
+			if len(infos) != 2 {
+				infos = strings.SplitN(infoPath, "..", 2) // match github behavior
+			}
+			ref2IsBranch := gitrepo.IsBranchExist(ctx, ctx.Repo.Repository, infos[1])
+			ref2IsTag := gitrepo.IsTagExist(ctx, ctx.Repo.Repository, infos[1])
```
Comment (wxiaoguang) [inline routers/web/repo/compare.go]: The path format is like this:

What if the path is `{:baseBranch}...{:headOwner}/{:headRepoName}:{:headBranch}.patch`?

image

Comment (badhezi) [reply routers/web/repo/compare.go]: Thanks ill address that

File: modules/git/tree.go (hunk 1 above)
Comment (badhezi) [reply modules/git/tree.go]: Thanks ill either move it to test files or find a workaround

File: routers/web/repo/compare.go [hunk 3]
```diff
@@ -265,7 +263,8 @@ func ParseCompareInfo(ctx *context.Context) *common.CompareInfo {
 				}
 				return nil
 			}
-			ci.HeadBranch = headInfos[1]
+			// FIXME: how to correctly choose the head repository? The logic below (3-8) is quite complex, the real head repo is determined there
```
Comment (wxiaoguang) [inline routers/web/repo/compare.go]: When trying to understand the logic, I find something strange, so I left a FIXME here.

Do you have ideas?

Comment (wxiaoguang) [reply routers/web/repo/compare.go]: ps: if it is too complex and we don't really need to support these cases, we can also skip the diff/patch support for these cases and leave some comments.

Comment (badhezi) [reply routers/web/repo/compare.go]: I'll address that, I prefer to support all use cases

Comment (badhezi) [reply routers/web/repo/compare.go]: @wxiaoguang is it clearer now?
I think i covered all use cases with integration tests as well, you think there is any other case we need to cover?

Comment (wxiaoguang) [reply routers/web/repo/compare.go]: I don't think so.

1. The new logic is inconsistent with the steps 3-8 below. The "headRepo" guess is quite complex.
2. I do not think GetRepoRawDiffForFile would work for different base&head repositories, for example: `ref1...headRepo:ref2`, but what if ref1 only exists in base repo and ref2 only exists in head repo?

File: routers/web/repo/compare.go [hunk 4]
```diff
@@ -743,6 +753,15 @@ func CompareDiff(ctx *context.Context) {
 		return
 	}
 
+	if ci.RawDiffType != "" {
+		err := git.GetRepoRawDiffForFile(ci.HeadGitRepo, ci.BaseBranch, ci.HeadBranch, ci.RawDiffType, "", ctx.Resp)
+		if err != nil {
+			ctx.ServerError("GetRepoRawDiffForFile", err)
+			return
```
Comment (lunny) [inline routers/web/repo/compare.go]:
```suggestion
```

File: tests/integration/compare_test.go [hunk 5]
```diff
@@ -158,3 +161,212 @@ func TestCompareCodeExpand(t *testing.T) {
...
+			Name:          "test_raw_diff",
+			Readme:        "Default",
+			AutoInit:      true,
+			DefaultBranch: "main",
+		}, true)
+		assert.NoError(t, err)
+		session := loginUser(t, user1.Name)
+
+		r, _ := gitrepo.OpenRepository(db.DefaultContext, repo)
+
```
Comment (lunny) [inline tests/integration/compare_test.go]:
```suggestion
defer r.Close()
```

**Comment by lunny** [comment]
If the from commit message is other languages with utf8. Is that right?

image

File: routers/web/repo/compare.go (hunk 3 above)
Comment (wxiaoguang) [reply routers/web/repo/compare.go]: ping @badhezi

**Comment by lunny** [comment]
Since #36105 merged, this PR needs some adjustments.

**Comment by wxiaoguang** [comment]
"add raw diff/patch endpoint for repository comparisons #37632" does the correct approach
